broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/NewPartitionTransitionImplTest.java

```diff
@@ -81,75 +80,90 @@ void shouldCallTransitionStepsInOrder() {
   }
 
   @Test
-  void shouldAbortTransitionIfOneStepThrowsAnException() {
+  // regression test for https://github.com/camunda-cloud/zeebe/issues/7873
+  void shouldNotStartMultipleTransitions() {
     // given
-    final var testException = new Exception("TEST_EXCEPTION");
-    when(mockStep1.transitionTo(mockContext, DEFAULT_TERM, DEFAULT_ROLE))
-        .thenReturn(TEST_CONCURRENCY_CONTROL.failedFuture(testException));
-    when(mockStep2.transitionTo(mockContext, DEFAULT_TERM, DEFAULT_ROLE))
+    final var firstStepFirstTransitionFuture = TEST_CONCURRENCY_CONTROL.<Void>createFuture();
+    final var firstStepSecondTransitionFuture = TEST_CONCURRENCY_CONTROL.<Void>createFuture();
+    final var firstStepThirdTransitionFuture = TEST_CONCURRENCY_CONTROL.<Void>createFuture();
+    when(mockStep1.transitionTo(any(), anyLong(), any()))
+        .thenReturn(firstStepFirstTransitionFuture)
+        .thenReturn(firstStepSecondTransitionFuture)
+        .thenReturn(firstStepThirdTransitionFuture);
+    when(mockStep1.prepareTransition(any(), anyLong(), any()))
         .thenReturn(TEST_CONCURRENCY_CONTROL.completedFuture(null));
 
-    final var sut = new NewPartitionTransitionImpl(of(mockStep1, mockStep2), mockContext);
-    sut.setConcurrencyControl(TEST_CONCURRENCY_CONTROL);
+    final var transition = new NewPartitionTransitionImpl(of(mockStep1), mockContext);
+    transition.setConcurrencyControl(TEST_CONCURRENCY_CONTROL);
+
+    final var firstTransitionFuture = transition.transitionTo(1, Role.FOLLOWER);
 
     // when
-    final var actualResult = sut.transitionTo(DEFAULT_TERM, DEFAULT_ROLE);
+    transition.transitionTo(2, Role.LEADER);
+    transition.transitionTo(2, Role.FOLLOWER);
```

Comment:
:question: Other dumb question :smile: we test later that we execute transition 1, which is good, then that we execute transition 2 with role LEADER. Why are we not executing transition 2 with role FOLLOWER again?

Reply:
Also, wouldn't we want to assert that the future returned by the other transitions is still completed regardless? We wouldn't want anything waiting on these forever, no?